source/_posts/2015-12-10-sometimes-its-good-to-be-shy.markdown

```diff
@@ -0,0 +1,242 @@
+
+In 1987, fruit of the work done on [The Demeter Project](http://www.ccs.neu.edu/research/demeter/), Ian Holland proposed a
+style rule for designing object-oriented systems. It was named "*[The Law of Demeter](http://www.ccs.neu.edu/home/lieber/LoD.html)*".
+What this rule defines, is a set of guidelines which restrict the amount
+of objects to which you may send messages to.
+
+I like [Brad Appleton's](http://www.bradapp.com/docs/demeter-intro.html) phrasing of the [style guideline](http://c2.com/cgi/wiki?LawOfDemeter):
+
+> A method "M" of an object "O" should invoke only the methods of the following kinds of objects:
+
+> 1. itself
+> 2. its parameters
+> 3. any objects it creates/instantiates
+> 4. its direct component objects
+
+Let's make it clearer with an example:
+
+> Choo choo! Here comes the train wreck!
+
+{% codeblock lang:ruby %}
+  def make_garden_pretty(gardener)
+    gardener.tools.garden_scissors.cut
+  end
+{% endcodeblock %}
+
+This small snippet, presents a method which violates Demeter's law. On
+the quest to make our garden beautiful, we are taking our chain of calls
+one step too far. Can you hear the train wreck coming?
+
+Demeter states on it's second rule that a method can invoke functions of the kind of object it
+receives as a parameter. However, upon the advent of a new object type, call
+chaining should stop. Our gardening example shows we're calling a method
+defined on the `Gardener` class, `tools`, which corresponds to the received
+parameter's type. However, chaining continues by invoking the `garden_scissors` method
+of the `Tool` module. Additionally, we're chaining the `cut` call, belonging to
+the `Task` module. Can you notice the problems this could lead to?
+
+Besides being kind of rude, this presents a bigger headache, very similar
+to the one depicted on the `PizzaBoy` examples. We are coupling
+ourselves to the outside world's concerns, creating a dependency which
+could ruin things for us in the future. Let me clarify this. Say some
+time from now, our gardener decides to stop using scissors in his daily
+work, and start using a lawn mower, because he determines that is the most
+effective way to make your garden shine. A change in his set of tools
+has developed. Scissors no longer exist in the gardener's toolbox.
+
+Knowing too much about how to make your garden look pretty lead to you
+breaking in response to a change in the `Gardener`'s implementation.
+In contrast to our `PizzaBoy` solution, we're not being shy here. In fact, we're
+not being shy **_at all_**.
+
+Picture this: Mr. gardener knocks on your door, you open it and
+he kindly salutes you. Immediately afterwards, as a result of wanting to
+make your garden prettier, you proceed to ask the gardener to hand you
+his set of tools, from which you grab a pair of scissors, and top it off
+by cutting your own grass. You might be asking yourself... Why did we call the
+gardener in the first place?
```

Comment:
I'd probably remove this entire paragraph since the point is remarked enough across the article. It's a little redundant to me.